optuna/trial/_base.py

```diff
@@ -52,6 +53,38 @@ def suggest_int(self, name: str, low: int, high: int, step: int = 1, log: bool =
 
         raise NotImplementedError
 
+    @overload
+    @abc.abstractmethod
+    def suggest_categorical(self, name: str, choices: Sequence[None]) -> None:
+        ...
+
+    @overload
+    @abc.abstractmethod
+    def suggest_categorical(self, name: str, choices: Sequence[bool]) -> bool:
+        ...
+
+    @overload
+    @abc.abstractmethod
+    def suggest_categorical(self, name: str, choices: Sequence[int]) -> int:
+        ...
+
+    @overload
+    @abc.abstractmethod
+    def suggest_categorical(self, name: str, choices: Sequence[float]) -> float:
+        ...
+
+    @overload
+    @abc.abstractmethod
+    def suggest_categorical(self, name: str, choices: Sequence[str]) -> str:
+        ...
+
+    @overload
+    @abc.abstractmethod
+    def suggest_categorical(
+        self, name: str, choices: Sequence[CategoricalChoiceType]
+    ) -> CategoricalChoiceType:
+        ...
+
     @abc.abstractmethod
     def suggest_categorical(
         self, name: str, choices: Sequence[CategoricalChoiceType]
```

Comment:
[Note] When I read [PEP 484](https://peps.python.org/pep-0484/#function-method-overloading), I thought we could remove the type hints here, but I confirmed that we needed them actually. The example code in the [mypy document](https://mypy.readthedocs.io/en/stable/more_types.html) also has the hints.